### PR TITLE
PHP 7.1 updates

### PR DESCRIPTION
**Pull request by asgrim**
Oops! I accidentally upgrade to minimum of PHP 7.1 :fearful: 

Everything should be "mostly" BC compat, but there will likely be some breaks if folks have been using internal stuff.

### CODE REVIEW

**Comment by Ocramius** [comment]
> Oops! I accidentally upgrade to minimum of PHP 7.1 😨

OOOOOOOPS! :trollface:

File: namespace-bc-aliases.php [hunk 1]
```diff
@@ -19,7 +19,6 @@ class_alias(\Roave\BetterReflection\Reflection\Exception\InvalidDefaultValueType
 class_alias(\Roave\BetterReflection\Reflection\Exception\MethodPrototypeNotFound::class, \BetterReflection\Reflection\Exception\MethodPrototypeNotFound::class);
 class_alias(\Roave\BetterReflection\Reflection\Exception\NoParent::class, \BetterReflection\Reflection\Exception\NoParent::class);
 class_alias(\Roave\BetterReflection\Reflection\Exception\NotAClassReflection::class, \BetterReflection\Reflection\Exception\NotAClassReflection::class);
-class_alias(\Roave\BetterReflection\Reflection\Exception\NotAString::class, \BetterReflection\Reflection\Exception\NotAString::class);
```
Comment (Ocramius) [inline namespace-bc-aliases.php]: This is a BC break that needs documentation - also, let's not be scared to call it a 2.0.0.

File: src/Identifier/Identifier.php [hunk 2]
```diff
@@ -14,7 +15,7 @@ class Identifier
      */
     private $type;
 
-    public function __construct($name, IdentifierType $type)
+    public function __construct(string $name, IdentifierType $type)
```
Comment (Ocramius) [inline src/Identifier/Identifier.php]: Kinda BC break (there was no docblock previously)

File: src/NodeCompiler/CompileNodeToValue.php [hunk 3]
```diff
@@ -77,7 +77,8 @@ private function compileArray(Node\Expr\Array_ $arrayNode, CompilerContext $cont
      * Compile constant expressions
      *
      * @param Node\Expr\ConstFetch $constNode
-     * @return bool|null
+     * @return bool|mixed|null
+     * @throws \Roave\BetterReflection\NodeCompiler\Exception\UnableToCompileNode
```
Comment (Ocramius) [inline src/NodeCompiler/CompileNodeToValue.php]: Just noticed in this code - this uses `constant()`, which may cause autoloading... new issue?

File: src/NodeCompiler/CompilerContext.php [hunk 4]
```diff
@@ -21,7 +21,7 @@ class CompilerContext
      * @param Reflector $reflector
      * @param ReflectionClass|null $self
      */
-    public function __construct(Reflector $reflector, ReflectionClass $self = null)
+    public function __construct(Reflector $reflector, ?ReflectionClass $self)
```
Comment (Ocramius) [inline src/NodeCompiler/CompilerContext.php]: This is a BC break due to optionality being removed - needs documenting

File: src/NodeCompiler/CompilerContext.php [hunk 5]
```diff
@@ -44,7 +44,7 @@ public function hasSelf()
      *
      * @return ReflectionClass|null
      */
```
Comment (Ocramius) [inline src/NodeCompiler/CompilerContext.php]: Can remove docblock

File: src/NodeCompiler/CompilerContext.php [hunk 6]
```diff
@@ -58,7 +58,7 @@ public function getSelf()
      *
      * @return Reflector
      */
```
Comment (Ocramius) [inline src/NodeCompiler/CompilerContext.php]: Docblock can be removed

File: src/Reflection/Exception/NotAString.php [hunk 7]
```diff
@@ -1,19 +0,0 @@
-<?php
-
-namespace Roave\BetterReflection\Reflection\Exception;
-
-class NotAString extends \InvalidArgumentException
```
Comment (Ocramius) [inline src/Reflection/Exception/NotAString.php]: This is a BC break that needs documenting

File: src/Reflection/Exception/NotAnInterfaceReflection.php [hunk 8]
```diff
@@ -11,7 +11,7 @@ class NotAnInterfaceReflection extends \UnexpectedValueException
      *
      * @return self
      */
```
Comment (Ocramius) [inline src/Reflection/Exception/NotAnInterfaceReflection.php]: Docblock can be removed

File: src/Reflection/ReflectionClass.php [hunk 9]
```diff
@@ -71,7 +70,7 @@ private function __construct()
      * @param string $className
      * @return string
      */
-    public static function export($className = null)
+    public static function export(string $className = null) : string
```
Comment (Ocramius) [inline src/Reflection/ReflectionClass.php]: is it `string` or `?string` here?

File: src/Reflection/ReflectionClass.php [hunk 10]
```diff
@@ -177,7 +176,7 @@ public function __toString()
      * @param string $className
      * @return ReflectionClass
      */
-    public static function createFromName($className)
+    public static function createFromName(string $className)
```
Comment (Ocramius) [inline src/Reflection/ReflectionClass.php]: Return type hint needed

File: src/Reflection/ReflectionClass.php [hunk 11]
```diff
@@ -190,10 +189,11 @@ public static function createFromName($className)
      * @see ReflectionObject::createFromInstance
      * @param object $instance
      * @return ReflectionClass
+     * @throws \InvalidArgumentException
      */
     public static function createFromInstance($instance)
     {
-        if (gettype($instance) !== 'object') {
+        if (!is_object($instance)) {
```
Comment (Ocramius) [inline src/Reflection/ReflectionClass.php]: Spacing

File: src/Reflection/ReflectionClass.php [hunk 12]
```diff
@@ -190,10 +189,11 @@ public static function createFromName($className)
      * @see ReflectionObject::createFromInstance
      * @param object $instance
      * @return ReflectionClass
+     * @throws \InvalidArgumentException
      */
     public static function createFromInstance($instance)
     {
-        if (gettype($instance) !== 'object') {
```
Comment (Ocramius) [inline src/Reflection/ReflectionClass.php]: Lol, how did we miss this :D

**Comment by asgrim** [comment]
bah, failures. Will need to investigate when I get some time, but I ran out of time :/

File: src/Reflection/ReflectionClass.php [hunk 13]
```diff
@@ -468,7 +468,7 @@ public function hasConstant($name)
      *
      * @return ReflectionMethod
      */
-    public function getConstructor()
+    public function getConstructor() : ReflectionMethod
```
Comment (Ocramius) [inline src/Reflection/ReflectionClass.php]: `?ReflectionMethod`? Or is this supposed to throw an exception?

File: src/Reflection/ReflectionClass.php [hunk 14]
```diff
@@ -468,7 +468,7 @@ public function hasConstant($name)
...
     {
         return $this->getMethod('__construct');
```
Comment (Ocramius) [inline src/Reflection/ReflectionClass.php]: TBH, could be php4 ctor... unsure though...

Comment (Ocramius) [inline src/Reflection/ReflectionClass.php]: php4 ctors only work with non-namespaced code (hint)

File: src/Reflection/ReflectionClass.php [hunk 15]
```diff
@@ -1097,7 +1089,7 @@ public function getStaticPropertyValue($propertyName)
      * @param mixed $value
      * @return void
      */
-    public function setStaticPropertyValue($propertyName, $value)
+    public function setStaticPropertyValue(string $propertyName, $value) : void
     {
         if (!class_exists($this->getName(), false)) {
```
Comment (Ocramius) [inline src/Reflection/ReflectionClass.php]: Just wondering whether we should load it or not (we have the located source...)

File: src/Reflection/ReflectionMethod.php [hunk 16]
```diff
@@ -245,7 +245,7 @@ public function isConstructor()
      *
      * @return bool
      */
-    public function isDestructor()
+    public function isDestructor() : bool
     {
         return $this->getName() === '__destruct';
```
Comment (Ocramius) [inline src/Reflection/ReflectionMethod.php]: Just noted that we don't really treat case **IN**-sensitivity correctly... :-\

File: src/Reflection/ReflectionObject.php [hunk 17]
```diff
@@ -151,103 +151,103 @@ private function createPropertyNodeFromReflection(\ReflectionProperty $property,
...
+    public function getImmediateMethods() : array
     {
         return $this->reflectionClass->getImmediateMethods();
     }
 
     /**
      * {@inheritdoc}
      */
-    public function getMethod($methodName)
+    public function getMethod(string $methodName) : ReflectionMethod
```
Comment (Ocramius) [inline src/Reflection/ReflectionObject.php]: Nullable?

File: src/Reflection/ReflectionProperty.php [hunk 18]
```diff
@@ -294,6 +295,7 @@ public function getDefaultValue()
 
     /**
      * {@inheritdoc}
+     * @throws \Roave\BetterReflection\Reflection\Exception\Uncloneable
```
Comment (Ocramius) [inline src/Reflection/ReflectionProperty.php]: Already imported this symbol

File: src/Reflection/ReflectionParameter.php [hunk 19]
```diff
@@ -538,6 +548,7 @@ public function getClass()
 
     /**
      * {@inheritdoc}
+     * @throws \Roave\BetterReflection\Reflection\Exception\Uncloneable
```
Comment (Ocramius) [inline src/Reflection/ReflectionParameter.php]: Already imported this symbol

File: src/Reflector/Reflector.php [hunk 20]
```diff
@@ -17,5 +17,5 @@
      * @return Reflection
      * @throws IdentifierNotFound
      */
-    public function reflect($identifierName);
+    public function reflect(string $identifierName) : Reflection;
```
Comment (Ocramius) [inline src/Reflector/Reflector.php]: Yeap, BC break to document...

File: src/SourceLocator/Ast/Exception/ParseToAstFailure.php [hunk 21]
```diff
@@ -11,7 +11,7 @@ class ParseToAstFailure extends \RuntimeException
      * @param \Exception|\Throwable $previous
```
Comment (Ocramius) [inline src/SourceLocator/Ast/Exception/ParseToAstFailure.php]: Just `\Throwable`

File: src/SourceLocator/Ast/Exception/ParseToAstFailure.php [hunk 22]
```diff
@@ -11,7 +11,7 @@ class ParseToAstFailure extends \RuntimeException
...
      * @return ParseToAstFailure
      */
-    public static function fromLocatedSource(LocatedSource $locatedSource, $previous = null)
+    public static function fromLocatedSource(LocatedSource $locatedSource, \Throwable $previous = null) : self
```
Comment (Ocramius) [inline src/SourceLocator/Ast/Exception/ParseToAstFailure.php]: Is the second parameter optional, or nullable?

File: src/SourceLocator/Exception/InvalidDirectory.php [hunk 23]
```diff
@@ -23,7 +23,7 @@ public static function fromNonDirectory($nonDirectory)
      *
      * @return InvalidDirectory
      */
-    public static function fromNonStringValue($nonStringValue)
+    public static function fromNonStringValue($nonStringValue) : self
```
Comment (Ocramius) [inline src/SourceLocator/Exception/InvalidDirectory.php]: Probably unused now, no?

File: src/SourceLocator/Type/AbstractSourceLocator.php [hunk 24]
```diff
@@ -27,7 +28,7 @@
      * @param Identifier $identifier
      * @return LocatedSource
```
Comment (Ocramius) [inline src/SourceLocator/Type/AbstractSourceLocator.php]: The docblock says not nullable - is this a BC break or a broken previous assumption?

File: src/SourceLocator/Type/AutoloadSourceLocator.php [hunk 25]
```diff
@@ -69,9 +73,9 @@ private function attemptAutoloadForIdentifier(Identifier $identifier)
...
      *
      * @param string $className
-     * @return string
+     * @return string|null
      */
-    private function locateClassByName($className)
+    private function locateClassByName(string $className)  : ?string
     {
         if (class_exists($className, false) || interface_exists($className, false) || trait_exists($className, false)) {
             return (new \ReflectionClass($className))->getFileName();
```
Comment (Ocramius) [inline src/SourceLocator/Type/AutoloadSourceLocator.php]: IIRC, this returns `string|bool` - to be checked

File: src/SourceLocator/Type/FileIteratorSourceLocator.php [hunk 26]
```diff
@@ -41,33 +42,32 @@ public function __construct(\Iterator $fileInfoIterator)
...
+    private function getAggregatedSourceLocator() : AggregateSourceLocator
     {
-        return $this->aggregateSourceLocator ?
-            $this->aggregateSourceLocator : new AggregateSourceLocator(array_values(array_filter(array_map(
-                function (\SplFileInfo $item) {
-                    if (! ($item->isFile() && pathinfo($item->getRealPath(), \PATHINFO_EXTENSION) === 'php')) {
-                        return null;
-                    }
+        return $this->aggregateSourceLocator ?: new AggregateSourceLocator(array_values(array_filter(array_map(
+            function (\SplFileInfo $item) {
```
Comment (Ocramius) [inline src/SourceLocator/Type/FileIteratorSourceLocator.php]: Return hint to be added

File: src/TypesFinder/FindTypeFromAst.php [hunk 27]
```diff
@@ -17,7 +18,7 @@ class FindTypeFromAst
      * @param string $namespace
      * @return \phpDocumentor\Reflection\Type|null
      */
-    public function __invoke($astType, LocatedSource $locatedSource, $namespace = '')
+    public function __invoke($astType, LocatedSource $locatedSource, string $namespace = '') : ?Type
```
Comment (Ocramius) [inline src/TypesFinder/FindTypeFromAst.php]: The `$astType` parameter type is not documented

File: src/Util/FindReflectionOnLine.php [hunk 28]
```diff
@@ -70,7 +74,7 @@ private function computeReflections($filename)
      * @return bool
      * @throws \InvalidArgumentException
      */
-    private function containsLine($reflection, $lineNumber)
+    private function containsLine($reflection, int $lineNumber) : bool
```
Comment (Ocramius) [inline src/Util/FindReflectionOnLine.php]: Is `$reflection` really `mixed`?

File: src/NodeCompiler/CompileNodeToValue.php [hunk 29]
```diff
@@ -260,9 +261,9 @@ private function compileBinaryOperator(Node\Expr\BinaryOp $node, CompilerContext
      * @param string          $constantName
      * @param ReflectionClass $class
      *
-     * @return CompilerContext|null
+     * @return ReflectionClass|null
```
Comment (TomasVotruba) [inline src/NodeCompiler/CompileNodeToValue.php]: These can be dropped now, all info is there.

File: src/Reflection/ReflectionClass.php (hunk 9 above)
Comment (Majkl578) [inline src/Reflection/ReflectionClass.php]: Both of course!
becausephp

File: src/NodeCompiler/CompileNodeToValue.php (hunk 3 above)
Comment (asgrim) [inline src/NodeCompiler/CompileNodeToValue.php]: Created #262

File: src/Reflection/ReflectionClass.php (hunk 15 above)
Comment (asgrim) [inline src/Reflection/ReflectionClass.php]: Side-effect surely?

File: src/Reflection/ReflectionClass.php (hunk 10 above)
Comment (asgrim) [inline src/Reflection/ReflectionClass.php]: Can't - `ReflectionObject` extends this, and ` : static` isn't a valid return type :/

Comment (asgrim) [inline src/Reflection/ReflectionClass.php]: And `ReflectionObject::createFromName` returns a `ReflectionObject` not a `ReflectionClass`, if that wasn't clear :)

File: src/Reflection/ReflectionClass.php (hunk 13 above)
Comment (asgrim) [inline src/Reflection/ReflectionClass.php]: Throws exception, have added doc

File: src/Reflection/ReflectionClass.php (hunk 14 above)
Comment (asgrim) [inline src/Reflection/ReflectionClass.php]: Created #263

File: src/Reflection/ReflectionMethod.php (hunk 16 above)
Comment (asgrim) [inline src/Reflection/ReflectionMethod.php]: Don't follow what you mean here?

File: src/Reflection/ReflectionObject.php (hunk 17 above)
Comment (asgrim) [inline src/Reflection/ReflectionObject.php]: No, throws exception (documented in `ReflectionClass` already)

File: src/SourceLocator/Ast/Exception/ParseToAstFailure.php (hunk 22 above)
Comment (asgrim) [inline src/SourceLocator/Ast/Exception/ParseToAstFailure.php]: optional

File: src/SourceLocator/Exception/InvalidDirectory.php (hunk 23 above)
Comment (asgrim) [inline src/SourceLocator/Exception/InvalidDirectory.php]: Nope, it's used in `DirectoriesSourceLocator` when iterating the `array` of `$directories` given, so still used

File: src/SourceLocator/Type/AbstractSourceLocator.php (hunk 24 above)
Comment (asgrim) [inline src/SourceLocator/Type/AbstractSourceLocator.php]: Fixing broken documentation, `createLocatedSource` has always been able to return `null` so doc was wrong

File: src/Reflection/ReflectionClass.php (hunk 15 above)
Comment (Ocramius) [inline src/Reflection/ReflectionClass.php]: Yup, but at this point the user probably wants that to happen anyway. Maybe we should store the value when we hijack the autoloading, and set it when the actual autoloading happens? Better than throwing an exception. This needs further discussion though.

File: src/Reflection/ReflectionMethod.php (hunk 16 above)
Comment (Ocramius) [inline src/Reflection/ReflectionMethod.php]: https://3v4l.org/i0QjV

```php
<?php

class Foo
{
    public function __dEsTrUcT() { var_dump(__METHOD__); }
}

new Foo;
```

File: src/Reflection/ReflectionObject.php (hunk 17 above)
Comment (Ocramius) [inline src/Reflection/ReflectionObject.php]: Alright

File: src/SourceLocator/Ast/Exception/ParseToAstFailure.php (hunk 22 above)
Comment (Ocramius) [inline src/SourceLocator/Ast/Exception/ParseToAstFailure.php]: 👍

File: src/SourceLocator/Exception/InvalidDirectory.php (hunk 23 above)
Comment (Ocramius) [inline src/SourceLocator/Exception/InvalidDirectory.php]: 👍

File: src/SourceLocator/Type/AbstractSourceLocator.php (hunk 24 above)
Comment (Ocramius) [inline src/SourceLocator/Type/AbstractSourceLocator.php]: 👍

File: src/Reflection/ReflectionMethod.php (hunk 16 above)
Comment (asgrim) [inline src/Reflection/ReflectionMethod.php]: Created #264 tho, if you type your destructors like that, you should be slapped with a fish :D

File: src/Reflection/ReflectionClass.php (hunk 15 above)
Comment (asgrim) [inline src/Reflection/ReflectionClass.php]: Created #265

File: src/NodeCompiler/CompileNodeToValue.php (hunk 29 above)
Comment (asgrim) [inline src/NodeCompiler/CompileNodeToValue.php]: Only reason I'd leave some of these in is where the method has an actual description (as IDEs can complain about missing doc info for params/return types even tho they're there). I'd rather keep them in until IDEs be a bit more clever about it :) but yes, this one I've removed.

**Comment by Ocramius** [comment]
🔫 BANG!

**Comment by TomasVotruba** [comment]
KA :boom:  ! 

:clap:  :clapper:

**Pull request merged**